### DISCUSSION TITLE
페이지교체 알고리즘, 프로세스와 스레드

### DISCUSSION THREAD

**Discussion by 24tngus** [24tngus/CS_STUDY#43] [Operating System]
#### 1. LFD (Longest Forward Distance)

- 1) 개념
  - 오프라인 알고리즘
  - 가장 먼 미래에 참조되는 페이지와 현재의 페이지를 바꾸는 알고리즘

- 2) 특징
  - 미래에 사용되는 프로세스 알지 못해서 사용X
  - 다른 알고리즘 성능 비교에대한 기준으로 사용
  
- 🗒️ 예시 : 프레임의 페이지 중 1이 뒤에 쓰일 것을 알고 미리 교체 
![](https://velog.velcdn.com/images/24tngus/post/38b7c308-173a-4e5f-a3d8-16719c4cb19d/image.png)

#### 2. FIFO (First In First Out)

- 1) 개념
  - 가장 먼저 온 페이지부터 교체
  
- 2) 특징
  - 구현 간단, 성능⬇️
  - 큐를 이용해 저장
  - Belady's Anomaly 현상 발생 가능 

- 🗒️ 예시 : 가장 오래된 7 교체
![](https://velog.velcdn.com/images/24tngus/post/d2c0352a-75f3-4c41-984f-253390f0b4c8/image.png)

  > 📖 참고 📖 Belady's Anomaly
  - 직관적으로, 프레임의 수가 많아질 수록 페이지 결함의 횟수는 감소함
  - 프레임 개수 많아져도 page-fault가 줄어들지 않고 늘어나는 현상
![](https://velog.velcdn.com/images/24tngus/post/97c94a94-9aed-49f4-a3b8-435138477736/image.png)
![](https://velog.velcdn.com/images/24tngus/post/317e5e52-34c7-4c7f-81eb-0d49231ebc08/image.png)

#### 3. LRU (Least Recently Used)

- 1) 개념
  - 가장 오래 사용하지 않은 페이지 교체 
  - 시간 지역성 성질 고려 (최근에 참조된 페이지가 가까운 미래에 다시 참조될 가능성 높음)
  
- 2) 특징
  - 성능 좋음
  - 최적 알고리즘과 비슷한 효과
  - 각 페이지마다 최근 사용한 시간 저장하는 구조 필요 (계수기, 스택)
  - 프로세스가 주기억장치에 접근할 때마다 참조된 페이지 시간을 기록해야함 ➡️ 막대한 오버헤드
- 🗒️ 예시 : 가장 오랫동안 사용하지 않은 1 교체
![](https://velog.velcdn.com/images/24tngus/post/a0f3c4cd-6a5e-4c55-9f38-01e8e1fc038e/image.png)

#### 4. NUR (Not Used Recently)

- 1) 개념
  - clock 알고리즘, LRU에서 발전된 알고리즘 
  - 최근에 사용하지 않은 페이지 교체 
  - 0과 1가진 비트 필요
    - 1 : 최근에 참조됨
    - 0 : 참조되지 않음
  ➡️ 시계 방향으로 돌면서 0을 찾으면 해당 페이지 1로 교체
  
- 2) 특징
  - 동일 그룹 내에서 선택 무작위 
  - 각 페이지마다 2개의 참조비트 / 변형 비트 사용
    - 페이지 프레임에 적재 : r = 0, m = 0
    - 페이지를 참조 : r = 1 (주기적으로 0으로 변경)
    - 페이지 수정 : m = 1
  - 교체 우선순위 : 참조비트 > 변형비트 
    - 1️⃣ : r = 0, m = 0
    - 2️⃣ : r = 0, m = 1
    - 3️⃣ : r = 1, m = 0
    - 4️⃣ : r = 1, m = 1

- 🗒️ 예시 : 우선순위 낮은 B 교체 
![](https://velog.velcdn.com/images/24tngus/post/267737f0-77c4-4fda-9e3c-a53ba58d3ecf/image.png)

#### 5. LFU (Least Frequently Used)

- 1) 개념
  - 가장 참조 횟수가 적은 페이지 교체
  - 교체 대상이 여러개라면 가장 오랫동안 사용하지 않은 페이지 교체 
  
- 2) 특징
  - LRU는 직전 참조된 시점만 반영
  - LFU는 참조 횟수를 통해 장기적으로 참조를 고려 가능
  ➡️ 막대한 오버헤드 

- 🗒️ 예시 : 각각 1번씩 참조한 페이지 2, 3 중 LRU에 따라 더 오래 쓰지 않은 2 교체
![](https://velog.velcdn.com/images/24tngus/post/9c42ad3f-41e7-418c-a9bf-725f156e18fa/image.png)

#### 1. process

- 1) 개념
  - 실행중인 프로그램
  - CPU 스케줄링 대상이 되는 작업(task)
  - 디스크로부터 메모리에 적재되어 CPU의 할당(주소 공간, 파일, 메모리 등) 받을 수 있음
  
- 2) 구성
  - 스택 (함수의 매개변수, 로컬변수 등 임시 자료 저장) 
  - 데이터 (전역변수 저장)
  - 힙 (동적 할당 메모리)
  - 코드
  ![](https://velog.velcdn.com/images/24tngus/post/2d35f908-9158-422e-a82b-f16da0c4f6d0/image.png)

- 3) prgram과 process 차이 
  - 프로그램이 메모리에 올라가면 프로세스가 되는 인스턴스화 발생
  - 운영체제의 CPU 스케줄러에 따라 CPU가 프로세스 실행
![](https://velog.velcdn.com/images/24tngus/post/273230ec-ad69-4fba-9afd-9efd2ac39a02/image.png)

#### 2. thread

- 1) 개념
  - 프로세스의 실행 단위
  - 한 프로세스 내에서 동작되는 실행 흐름, 프로세스 내의 주소 공간이나 자원 공유

- 2) 구성
  - 스레드 ID
  - 프로그램 카운터 : 스레드가 스케줄러에 의해 CPU 할당 받을 때 명령어 어디까지 수행했는지 저장
  ➡️ PC 레지스터 독립적 할당
  - 레지스터 집합
  - 스택 : 함수 호출시 전달 인자, 되돌아갈 주소값, 함수 내 선언 변수 저장
  ➡️ 독립적인 함수 호출 등 실행을 위해 스택 독립적 할당

- 3) 특징
  - **동시성** (서로 독립적인 작업들을 작은 단위로 나누고, 동시에 실행되는 것처럼 보여줌)
  - 자원의 생성과 관리 중복성 최소화 
  - 한 스레드가 중단(block)되어도 다른 스레드는 실행(runnig) 상태 일수 있어 중단되지 않고 빠르게 처리
  
- 4) 구분
  - 싱글스레드 프로세스 : 하나의 스레드로 이루어진 프로세스
  - 멀티스레드 프로세스 : 여러개의 스레드로 이루어진 프로세스
![](https://velog.velcdn.com/images/24tngus/post/77c331fc-1846-43bb-b530-6ec82238efce/image.png)

#### 3. process와 thread 차이점

- 1) 작업 메모리 영역
  - process : 코드, 데이터, 스택, 힙 메모리 영역
  - thread : 스택 제외한 다른 메모리 영역을 스레들과 공유
- 2) 프로세스간 통신
  - process : 타 프로세스와 격리되어 있어 IPC 사용 
  - thread : 타 스레드와 연결되어 있어 정보 공유 쉽지만, 동기화 문제점 
- 3) 프로세스간 영향
  - process : 프로세스 문제가 타 프로세스에 영향X 
  - thread : 한 스레드에 문제시 타 스레드에 영향 미침
- 4) 생성과 종료
  - process : 많은 시간 소요 
  - thread : 적은 시간 소요
![](https://velog.velcdn.com/images/24tngus/post/b6bd1ba1-e151-400f-a507-d404f54260e8/image.png)

**Comment by jindream6128** [comment]
정처기가 떠오릅니다.. ㅎ

**Comment by berryberries** [comment]
확인했습니다!
